fe/fe-core/src/main/java/com/starrocks/load/routineload/PulsarProgress.java

```diff
@@ -0,0 +1,161 @@
+    private Map<String, Long> partitionToBacklogNum = Maps.newConcurrentMap();
+    // Initial positions will only be used at first schedule
+    private Map<String, Long> partitionToInitialPosition = Maps.newConcurrentMap();
+
+    public PulsarProgress() {
+        super(LoadDataSourceType.PULSAR);
+    }
+
+    public PulsarProgress(TPulsarRLTaskProgress tPulsarRLTaskProgress) {
+        super(LoadDataSourceType.PULSAR);
+        this.partitionToBacklogNum = tPulsarRLTaskProgress.getPartitionBacklogNum();
+    }
+
+    public Map<String, Long> getPartitionToInitialPosition(List<String> partitions) {
+        Map<String, Long> result = Maps.newHashMap();
+        for (Map.Entry<String, Long> entry : partitionToInitialPosition.entrySet()) {
+            for (String partition : partitions) {
+                if (entry.getKey().equals(partition)) {
+                    result.put(partition, entry.getValue());
+                }
+            }
+        }
+        return result;
+    }
+
+    public List<Long> getBacklogNums() {
+        return new ArrayList<Long>(partitionToBacklogNum.values());
+    }
+
+    public Long getInitialPosition(String partition) {
+        if (partitionToInitialPosition.containsKey(partition)) {
+            return partitionToInitialPosition.get(partition);
+        } else {
+            return -1L;
+        }
+    }
+
+    public void addPartitionToInitialPosition(Pair<String, Long> partitionToInitialPosition) {
+        this.partitionToInitialPosition.put(partitionToInitialPosition.first, partitionToInitialPosition.second);
+    }
+
+    public void modifyInitialPositions(List<Pair<String, Long>> partitionInitialPositions) {
+        for (Pair<String, Long> pair : partitionInitialPositions) {
+            this.partitionToInitialPosition.put(pair.first, pair.second);
+        }
+    }
+
+    public void unprotectUpdate(List<String> currentPartitions, Long defaultInitialPosition) {
+        partitionToInitialPosition.keySet().stream()
+                .filter(entry -> !currentPartitions.contains(entry))
+                .forEach(entry -> partitionToInitialPosition.remove(entry));
+
+        if (defaultInitialPosition != null) {
+            currentPartitions.stream()
+                    .filter(entry -> !partitionToInitialPosition.containsKey(entry))
+                    .forEach(entry -> partitionToInitialPosition.put(entry, defaultInitialPosition));
+        }
+    }
+
+    private void getReadableProgress(Map<String, String> showPartitionIdToPosition) {
+        for (Map.Entry<String, Long> entry : partitionToBacklogNum.entrySet()) {
+            showPartitionIdToPosition.put(entry.getKey() + "(BacklogNum)", String.valueOf(entry.getValue()));
+        }
+    }
+
+    @Override
+    public String toString() {
+        Map<String, String> showPartitionToBacklogNum = Maps.newHashMap();
+        getReadableProgress(showPartitionToBacklogNum);
+        return "PulsarProgress [partitionToBacklogNum="
```

Comment:
Why is the position not included here?

Reply:
Progress should keep 2 kinds of info:

1. Consume progress;
2. Consume position for exactly-once semantic.

Kafka uses `partitionIdToOffset` for both purpose.

Pulsar only needs backlog num for consume progress record here. Let's support exactly-once semantic later.